<a href="https://colab.research.google.com/github/dpenamoya-ops/TelecomX_LATAM_parte2/blob/main/TelecomX_LATAM_parte2_dp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://trello.com/b/O1cSUxni/telecom-x-analisis-de-evasion-de-clientes-parte-2-dpena

#🛠️ Preparación de los Datos

##Extracción del Archivo Tratado

In [33]:
import pandas as pd

# Cargar el DataFrame
df = pd.read_csv("https://raw.githubusercontent.com/dpenamoya-ops/TelecomX_LATAM_parte2/main/datos_tratados.csv")

# Mostrar las primeras filas para verificar la carga
df.head(), df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7043 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


(   customerID Churn                                           customer  \
 0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
 1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
 2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
 3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
 4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
 
                                              phone  \
 0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
 1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
 2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
 3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
 4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
 
                                             internet  \
 0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
 1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
 2  {'InternetService': 'Fiber optic', 'Onlin

##Eliminación de Columnas Irrelevantes

In [34]:
#Revisar las columnas:
df.columns

Index(['customerID', 'Churn', 'customer', 'phone', 'internet', 'account'], dtype='object')

In [35]:
#identificar posibles columnas ID
df.nunique().sort_values(ascending=False).head(10)

customerID    7267
account       6931
customer       891
internet       129
phone            3
Churn            2
dtype: int64

In [36]:
# Eliminar columnas irrelevantes
df = df.drop(columns=['customerID'])

# Verificar resultado
df.head()

Churn                                           customer  \
0    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

evaluamos columnas account, account, internet

In [37]:
df['account'].head()

0    {'Contract': 'One year', 'PaperlessBilling': '...
1    {'Contract': 'Month-to-month', 'PaperlessBilli...
2    {'Contract': 'Month-to-month', 'PaperlessBilli...
3    {'Contract': 'Month-to-month', 'PaperlessBilli...
4    {'Contract': 'Month-to-month', 'PaperlessBilli...
Name: account, dtype: object

In [38]:
df['customer'].value_counts().head()

customer
{'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 1}         223
{'gender': 'Female', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 1}       192
{'gender': 'Male', 'SeniorCitizen': 0, 'Partner': 'Yes', 'Dependents': 'Yes', 'tenure': 72}       77
{'gender': 'Female', 'SeniorCitizen': 0, 'Partner': 'Yes', 'Dependents': 'Yes', 'tenure': 72}     76
{'gender': 'Female', 'SeniorCitizen': 0, 'Partner': 'No', 'Dependents': 'No', 'tenure': 2}        74
Name: count, dtype: int64

In [39]:
df['internet'].value_counts().head()

internet
{'InternetService': 'No', 'OnlineSecurity': 'No internet service', 'OnlineBackup': 'No internet service', 'DeviceProtection': 'No internet service', 'TechSupport': 'No internet service', 'StreamingTV': 'No internet service', 'StreamingMovies': 'No internet service'}    1581
{'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'No', 'StreamingTV': 'No', 'StreamingMovies': 'No'}                                                                                                  417
{'InternetService': 'DSL', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'No', 'StreamingTV': 'No', 'StreamingMovies': 'No'}                                                                                                          304
{'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'No', 'DeviceProtection': 'No', 'TechSupport': 'No', 'StreamingTV': 'Yes', 'StreamingMovies': 'Yes'}                                                                                                186
{'InternetService': 'Fiber optic', 'OnlineSecurity': 'No', 'OnlineBackup': 'Yes', 'DeviceProtection': 'Yes', 'TechSupport': 'No', 'StreamingTV': 'Yes', 'StreamingMovies': 'Yes'}                                                                                              172
Name: count, dtype: int64

In [40]:
#confirmamos si son strings
type(df['account'][0])

str

In [41]:
#convertimos a diccionario real
import ast

df['account'] = df['account'].apply(ast.literal_eval)
df['customer'] = df['customer'].apply(ast.literal_eval)
df['internet'] = df['internet'].apply(ast.literal_eval)
df['phone'] = df['phone'].apply(ast.literal_eval)

In [42]:
#Expandirmos a columnas normales
df_account = df['account'].apply(pd.Series)
df_customer = df['customer'].apply(pd.Series)
df_internet = df['internet'].apply(pd.Series)
df_phone = df['phone'].apply(pd.Series)

In [43]:
#Unimos todo al dataframe original
df = pd.concat(
    [df.drop(columns=['account','customer','internet','phone']),
     df_account,
     df_customer,
     df_internet,
     df_phone],
    axis=1
)

In [44]:
df.head()

Churn        Contract PaperlessBilling     PaymentMethod  \
0    No        One year              Yes      Mailed check   
1    No  Month-to-month               No      Mailed check   
2   Yes  Month-to-month              Yes  Electronic check   
3   Yes  Month-to-month              Yes  Electronic check   
4   Yes  Month-to-month              Yes      Mailed check   

                                 Charges  gender  SeniorCitizen Partner  \
0    {'Monthly': 65.6, 'Total': '593.3'}  Female              0     Yes   
1    {'Monthly': 59.9, 'Total': '542.4'}    Male              0      No   
2   {'Monthly': 73.9, 'Total': '280.85'}    Male              0      No   
3  {'Monthly': 98.0, 'Total': '1237.85'}    Male              1     Yes   
4    {'Monthly': 83.9, 'Total': '267.4'}  Female              1     Yes   

  Dependents  tenure InternetService OnlineSecurity OnlineBackup  \
0        Yes       9             DSL             No          Yes   
1         No       9             DSL             No           No   
2         No       4     Fiber optic             No           No   
3         No      13     Fiber optic             No          Yes   
4         No       3     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies PhoneService  \
0               No         Yes         Yes              No          Yes   
1               No          No          No             Yes          Yes   
2              Yes          No          No              No          Yes   
3              Yes          No         Yes             Yes          Yes   
4               No         Yes         Yes              No          Yes   

  MultipleLines  
0            No  
1           Yes  
2            No  
3            No  
4            No

In [45]:
type(df['Charges'][0])

dict

In [46]:
charges_df = df['Charges'].apply(pd.Series)

In [47]:
charges_df.head()

Monthly    Total
0     65.6    593.3
1     59.9    542.4
2     73.9   280.85
3     98.0  1237.85
4     83.9    267.4

In [48]:
#convertimos a numérico
charges_df['Monthly'] = pd.to_numeric(charges_df['Monthly'], errors='coerce')
charges_df['Total'] = pd.to_numeric(charges_df['Total'], errors='coerce')

In [49]:
#renombramos columnas (más profesional)
charges_df = charges_df.rename(columns={
    'Monthly': 'MonthlyCharges',
    'Total': 'TotalCharges'
})

In [50]:
#integramos al df principal
df = pd.concat(
    [df.drop(columns=['Charges']), charges_df],
    axis=1
)

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Churn             7043 non-null   object 
 1   Contract          7267 non-null   object 
 2   PaperlessBilling  7267 non-null   object 
 3   PaymentMethod     7267 non-null   object 
 4   gender            7267 non-null   object 
 5   SeniorCitizen     7267 non-null   int64  
 6   Partner           7267 non-null   object 
 7   Dependents        7267 non-null   object 
 8   tenure            7267 non-null   int64  
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  PhoneService      7267 non-null   object 


In [52]:
#revisamos si se generaron NaN
df[['MonthlyCharges','TotalCharges']].isnull().sum()

MonthlyCharges     0
TotalCharges      11
dtype: int64

In [53]:
#confirmamos ternure
df[df['TotalCharges'].isnull()][['tenure','MonthlyCharges','TotalCharges']]

tenure  MonthlyCharges  TotalCharges
975        0           56.05           NaN
1775       0           20.00           NaN
1955       0           61.90           NaN
2075       0           19.70           NaN
2232       0           20.25           NaN
2308       0           25.35           NaN
2930       0           73.35           NaN
3134       0           25.75           NaN
3203       0           52.55           NaN
4169       0           80.85           NaN
5599       0           19.85           NaN

In [54]:
#reemplazamos los Nan por 0
df['TotalCharges'] = df['TotalCharges'].fillna(0)

In [55]:
#verificamos
df.isnull().sum()

Churn               224
Contract              0
PaperlessBilling      0
PaymentMethod         0
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
PhoneService          0
MultipleLines         0
MonthlyCharges        0
TotalCharges          0
dtype: int64

Observamos que la columna Churn aun tiene 224 valores NaN

In [56]:
df[df['Churn'].isnull()].head()

Churn        Contract PaperlessBilling              PaymentMethod  gender  \
30    NaN  Month-to-month              Yes  Bank transfer (automatic)  Female   
75    NaN        Two year              Yes    Credit card (automatic)    Male   
96    NaN        One year              Yes           Electronic check    Male   
98    NaN  Month-to-month               No    Credit card (automatic)  Female   
175   NaN        One year              Yes  Bank transfer (automatic)    Male   

     SeniorCitizen Partner Dependents  tenure InternetService  \
30               0      No         No      11     Fiber optic   
75               0      No         No      71              No   
96               0      No         No      29             DSL   
98               1      No         No       5             DSL   
175              1     Yes         No      65     Fiber optic   

          OnlineSecurity         OnlineBackup     DeviceProtection  \
30                   Yes                   No                   No   
75   No internet service  No internet service  No internet service   
96                   Yes                  Yes                   No   
98                   Yes                  Yes                   No   
175                   No                  Yes                  Yes   

             TechSupport          StreamingTV      StreamingMovies  \
30                    No                   No                   No   
75   No internet service  No internet service  No internet service   
96                   Yes                   No                   No   
98                   Yes                   No                   No   
175                   No                  Yes                  Yes   

    PhoneService MultipleLines  MonthlyCharges  TotalCharges  
30           Yes           Yes           79.00        929.30  
75           Yes            No           19.90       1355.10  
96           Yes            No           58.75       1696.20  
98           Yes            No           59.90        287.85  
175          Yes           Yes          103.15       6792.45

In [57]:
#eliminanos valores NaN
df = df.dropna(subset=['Churn'])

In [58]:
#verificamos que no quede ninguno
df['Churn'].isnull().sum()

np.int64(0)

Verificamos el  balance de clases

In [59]:
df['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

##Verificación de la Proporción de Cancelación (Churn)

In [60]:
df['Churn'].value_counts(normalize=True)

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

Interpretación del desbalance

Tenemos:

🔵 73.5% clientes que NO cancelan

🔴 26.5% clientes que SÍ cancelan

**¿Hay desbalance?**

Sí, pero no es extremo.

En machine learning:

50/50 → balance perfecto

70/30 → desbalance moderado ✅

90/10 → desbalance serio ⚠️

Este caso es un desbalance moderado.

**¿Por qué importa esto?**

Si entrenamos un modelo sin considerar esto, podría:

Aprender a predecir siempre "No"

Obtener 73% de accuracy

Pero ser inútil detectando churn real

Por eso más adelante deberemos:

Evaluar precision, recall y F1

No confiar solo en accuracy

Considerar class_weight='balanced' si usamos modelos como LogisticRegression

**Conclusión:** La proporción de cancelación es aproximadamente 26.5%, mientras que el 73.5% de los clientes permanecen activos. Existe un desbalance moderado entre las clases, lo cual podría afectar el rendimiento de modelos predictivos si no se consideran métricas adecuadas o técnicas de ajuste.

##Encoding
Transforma las variables categóricas a formato numérico para hacerlas compatibles con los algoritmos de machine learning. Utiliza un método de codificación adecuado, como one-hot encoding.

In [61]:
#Convertimos la variable objetivo
df['Churn'] = df['Churn'].map({'Yes':1, 'No':0})

In [62]:
df['Churn'].value_counts()

Churn
0    5174
1    1869
Name: count, dtype: int64

In [63]:
#Separamos X e y
X = df.drop('Churn', axis=1)
y = df['Churn']

In [64]:
#identificamos variables categoricas
categorical_cols = X.select_dtypes(include=['object']).columns
categorical_cols

Index(['Contract', 'PaperlessBilling', 'PaymentMethod', 'gender', 'Partner',
       'Dependents', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'PhoneService', 'MultipleLines'],
      dtype='object')

In [65]:
#aplicamos one hot encoding
X_encoded = pd.get_dummies(
    X,
    columns=categorical_cols,
    drop_first=True
)

In [66]:
#verificamos
X_encoded.head()

SeniorCitizen  tenure  MonthlyCharges  TotalCharges  Contract_One year  \
0              0       9            65.6        593.30               True   
1              0       9            59.9        542.40              False   
2              0       4            73.9        280.85              False   
3              1      13            98.0       1237.85              False   
4              1       3            83.9        267.40              False   

   Contract_Two year  PaperlessBilling_Yes  \
0              False                  True   
1              False                 False   
2              False                  True   
3              False                  True   
4              False                  True   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                  False                           False   
1                                  False                           False   
2                                  False                            True   
3                                  False                            True   
4                                  False                           False   

   PaymentMethod_Mailed check  ...  DeviceProtection_Yes  \
0                        True  ...                 False   
1                        True  ...                 False   
2                       False  ...                  True   
3                       False  ...                  True   
4                        True  ...                 False   

   TechSupport_No internet service  TechSupport_Yes  \
0                            False             True   
1                            False            False   
2                            False            False   
3                            False            False   
4                            False             True   

   StreamingTV_No internet service  StreamingTV_Yes  \
0                            False             True   
1                            False            False   
2                            False            False   
3                            False             True   
4                            False             True   

   StreamingMovies_No internet service  StreamingMovies_Yes  PhoneService_Yes  \
0                                False                False              True   
1                                False                 True              True   
2                                False                False              True   
3                                False                 True              True   
4                                False                False              True   

   MultipleLines_No phone service  MultipleLines_Yes  
0                           False              False  
1                           False               True  
2                           False              False  
3                           False              False  
4                           False              False  

[5 rows x 30 columns]

In [67]:
X_encoded.dtypes.value_counts()

bool       26
int64       2
float64     2
Name: count, dtype: int64

In [68]:
#transformamos los datos bool a binario
X_encoded = X_encoded.astype(int)

In [69]:
X_encoded.dtypes.value_counts()

int64    30
Name: count, dtype: int64

Hacemos el split/test obligatorio

In [70]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_encoded,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [71]:
y_train.value_counts(normalize=True)

Churn
0    0.734647
1    0.265353
Name: proportion, dtype: float64

##Balanceo de Clases
Si deseas profundizar en el análisis, aplica técnicas de balanceo como undersampling o oversampling. En situaciones de fuerte desbalanceo, herramientas como SMOTE pueden ser útiles para generar ejemplos sintéticos de la clase minoritaria.

Probamos los tres métodos: Undersampling, Oversampling y SMOTE

In [72]:
#Undersampling
from imblearn.under_sampling import RandomUnderSampler

rus = RandomUnderSampler(random_state=42)
X_train_under, y_train_under = rus.fit_resample(X_train, y_train)

In [76]:
y_train_under.value_counts()


Churn
0    1495
1    1495
Name: count, dtype: int64

In [73]:
#Oversampling
from imblearn.over_sampling import RandomOverSampler

ros = RandomOverSampler(random_state=42)
X_train_over, y_train_over = ros.fit_resample(X_train, y_train)

In [77]:
y_train_over.value_counts()


Churn
0    4139
1    4139
Name: count, dtype: int64

In [74]:
#SMOTE
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)

In [78]:
y_train_smote.value_counts()

Churn
0    4139
1    4139
Name: count, dtype: int64

##Normalización o Estandarización (si es necesario)
Evalúa la necesidad de normalizar o estandarizar los datos, según los modelos que se aplicarán. Modelos basados en distancia, como KNN, SVM, Regresión Logística y Redes Neuronales, requieren este preprocesamiento. Por otro lado, modelos basados en árboles, como Decision Tree, Random Forest y XGBoost, no son sensibles a la escala de los datos.

In [80]:
#Paso 1 — Importar StandardScaler
from sklearn.preprocessing import StandardScaler

In [83]:
#Paso 2 — Crear el escalador
scaler = StandardScaler()

In [84]:
#Paso 3 — Ajustar y transformar SOLO el entrenamiento balanceado
X_train_scaled = scaler.fit_transform(X_train_smote)

In [85]:
#Paso 4 — Transformar el test (sin hacer fit)
X_test_scaled = scaler.transform(X_test)

¿Por qué no hacemos fit en el test?

Porque eso sería data leakage.

El modelo debe "aprender" la media y desviación solo del entrenamiento

#🎯 Correlación y Selección de Variables